core/services/chainlink/relayer_chain_interoperators.go

```diff
@@ -0,0 +1,393 @@
+		}
+		err2 = op.putBatch(starkRelayers)
+		if err2 != nil {
+			return fmt.Errorf("failed to store StarkNet relayers: %w", err2)
+		}
+		return nil
+	}
+}
+
+func (rs *CoreRelayerChainInteroperators) Get(id relay.Identifier) (loop.Relayer, error) {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	lr, exist := rs.loopRelayers[id]
+	if !exist {
+		return nil, fmt.Errorf("%w: %s", ErrNoSuchRelayer, id)
+	}
+	return lr, nil
+}
+
+func (rs *CoreRelayerChainInteroperators) putOne(id relay.Identifier, lr loop.Relayer) error {
+
+	// backward compatibility. this is bit gross to type cast but it hides the details from products.
+	switch id.Network {
+	case relay.EVM:
+		adapter, ok := lr.(evmrelayer.LoopRelayAdapter)
+		if !ok {
+			return fmt.Errorf("unsupported evm loop relayer implementation. got %t want (evmrelayer.LoopRelayAdapter)", lr)
+		}
+
+		rs.legacyChains.EVMChains.Put(id.ChainID.String(), adapter.Chain())
+		if adapter.Default() {
+			dflt, _ := rs.legacyChains.EVMChains.Default()
+			if dflt != nil {
+				return fmt.Errorf("multiple default evm chains. %s, %s", dflt.ID(), adapter.Chain().ID())
+			}
+			rs.legacyChains.EVMChains.SetDefault(adapter.Chain())
+		}
+		rs.srvs = append(rs.srvs, adapter)
+	case relay.Cosmos:
+		adapter, ok := lr.(cosmos.LoopRelayerChainer)
+		if !ok {
+			return fmt.Errorf("unsupported cosmos loop relayer implementation. got %t want (cosmos.LoopRelayAdapter)", lr)
+		}
+
+		rs.legacyChains.CosmosChains.Put(id.ChainID.String(), adapter.Chain())
+		rs.srvs = append(rs.srvs, adapter)
+	default:
+		rs.srvs = append(rs.srvs, lr)
+	}
+
+	rs.loopRelayers[id] = lr
+	return nil
+}
+
+func (rs *CoreRelayerChainInteroperators) Put(id relay.Identifier, lr loop.Relayer) error {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	return rs.putOne(id, lr)
+}
+
+// TODO maybe make Relayer[U,V] where u,v are the chain specific types and then make this generic
+// Relayer[U evm.LoopAdapter, Vcosmos.LoopAdapter]
+// (rs Relayer[U,V] putBatch(map[](loop.relayer|U|V))
+func (rs *CoreRelayerChainInteroperators) putBatch(b map[relay.Identifier]loop.Relayer) (err error) {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	for id, r := range b {
+		err2 := rs.putOne(id, r)
+		if err2 != nil {
+			err = multierror.Append(err, err2)
+		}
+	}
+	return err
+}
+
+func (rs *CoreRelayerChainInteroperators) LegacyEVMChains() evm.LegacyChainContainer {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	return rs.legacyChains.EVMChains
+}
+
+func (rs *CoreRelayerChainInteroperators) LegacyCosmosChains() cosmos.LegacyChainContainer {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	return rs.legacyChains.CosmosChains
+}
+
+// ChainStatus gets [types.ChainStatus] relayID must be string representation of [relayer.Identifier], which ensures unique identification
+// amongst the multiple relayer:chain pairs wrapped in the interoperators
+func (rs *CoreRelayerChainInteroperators) ChainStatus(ctx context.Context, relayerID string) (types.ChainStatus, error) {
```

Comment:
Why don't we use the `relayer.Identifier` type here? Otherwise the signature clashes with the old approach of passing only chain ID.
```suggestion
func (rs *CoreRelayerChainInteroperators) ChainStatus(ctx context.Context, id relayer.Identifier) (types.ChainStatus, error) {
```

Reply:
Hmm but why are we now required to pass in something with more information?
It used to be just chain ID, but now that should be entirely implicit, no? Why do callers also need to include the network?

Reply:
yeah, good question. the answer is that this houses all relayers, so instead of chainset.ChainXXX.ChainStatus(chainid string) we need to identify which relayer we want.

this is a temporary state and will be changed in the follow up.

Reply:
> this is a temporary state and will be changed in the follow up.

nbd but I think I'm missing something still. If this method is for a different set of callers which pass different information in the string field, then why are we bound to the same interface? Shouldn't we be able to separate the two cases?

Reply:
you are right and i was confused. i was recycling a common interface for convenience and didn't have to implement it. updated to take relay.Identifier + comment about why it's different from the other interface